Change selected nodes with the keyboard in the flamegraph

brisad opened the pull request:
Taken from #1311

brisad commented on src/test/components/FlameGraph.test.js:
```diff
@@ -5,66 +5,115 @@
+describe('FlameGraph', function() {
+  function setup() {
+    const flushRafCalls = mockRaf();
+    window.devicePixelRatio = 1;
```
I realize I forgot to put back `delete window.devicePixelRatio`.

brisad commented:
Ah, yes it is. Adding you then :)

julienw commented:
Thanks for this patch, this is exciting :)

I had a very quick look at the behavior, not at the code yet, here are my small comments. I specifically tested with this profile: https://deploy-preview-1711--perf-html.netlify.com/public/671cf336e06a20333fbbfa599c76e0f496410cc0/flame-graph/?globalTrackOrder=0-1-2-3-4-5-6-7-8-9-10&hiddenGlobalTracks=1-2-3-4-5-7-8&thread=11&v=3

* when there's nothing selected yet, we should select "root" when the first key is pressed
* the "focus" style is not visible enough. Also we shouldn't use colors only. Maybe we can put a thicker border for these nodes?
* it's also painful to see the focus when the nodes are very small. We should likely skip the ones that are barely noticeable (have a width threshold). Also maybe thickening the border as advised above could make it better.

I'll look at the code tomorrow!

brisad commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -84,6 +84,61 @@ class FlameGraph extends React.PureComponent<Props> {
+            threadIndex,
+            getCallNodePathFromIndex(callNodeIndex, callNodeTable)
+          );
+        }
```
I don't really like that this code is using different data structures in order to navigate in different directions. `callNodeTable` up, `callTree` down, and `flameGraphTiming` and `callNodeTable` sideways. I initially tried to encapsulate this in `flamegraph.js`, but only working on `flameGraphTiming` would be inefficient since it would have to do a lot of search.
So I left it like this...

julienw replied:
I think it's good in the component actually :)

julienw commented on src/test/components/FlameGraph.test.js:
```diff
@@ -5,66 +5,114 @@
 
-  expect(container.firstChild).toMatchSnapshot();
+    expect(selectedNode()).toBe('B');
+  });
```
nit: I'd remove the empty lines for groups where this makes sense (eg: fireEvent something + the matching expectation). Also adding some comments for each of the "groups" explaining the expectations

julienw commented on src/test/components/FlameGraph.test.js:
```diff
@@ -5,66 +5,114 @@
 
-it('renders FlameGraph correctly', () => {
-  const flushRafCalls = mockRaf();
-  window.devicePixelRatio = 1;
```
I don't see this `window.devicePixelRatio` anywhere in the new code, is that expected?

brisad replied:
I removed it because it wasn't used (or so I believed) and the test ran fine without it.

julienw replied:
It's used in https://github.com/firefox-devtools/profiler/blob/423cfb5dc51809b0d1eabcc5cbab30997cc9d532/src/components/shared/chart/Canvas.js#L97-L99

But I see it's been added in jsdom 11.6 (we upgraded to 13 recently), so that's the reason ir works!

brisad replied:
Thanks for explaining why it worked :-)

julienw commented:
(and I forgot: thank you!)

brisad commented:
Thanks for the comments! I think I'll probably handle the UX comments in this same PR. Been quite some busy days lately, but I hope I'll get around to it soon.

brisad commented:
Hi @julienw ! I addressed two of your UX comments in the latest commit, namely selecting the root element and skipping boxes that are too narrow.

The width threshold i picked worked fine on my screen, but perhaps it will less than ideal for higher resolution screens, I don't know. Could you please take a look and see what you think? Is my arbitrary threshold okay? Things will probably get a bit more complicated if we need to dynamically find this threshold. I'll wait with implementing tests for this until I see what comments you have on this approach :-)

Addressing the comment you had about re-styling the focused node affects more than the keyboard navigation and also the stack chart since the styling looks the same there. So I think that one could actually be handled in a separate issue. What do you think? I'd be happy to write that issue.

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -116,7 +171,12 @@ class FlameGraph extends React.PureComponent<Props> {
+        const children = callTree.getChildren(selectedCallNodeIndex);
+        // The order of the children returned from getChildren does
+        // not correspond to the paint order in the flame graph, so
+        // the one chosen for selection might look rather arbitrary.
```
Actually I don't think this is so arbitrary: I believe the children are sorted by running time, so it actually looks quite logical to me.
This also means that the first children will be the widest of all children, so you can just test it, instead of using `find`, to know if it's possible to move up.

brisad replied:
Yes you're right. I was thinking about when the user navigates with the keyboard. If he or she doesn't realize that it is the largest box that is selected when going up, it might be confusing and look arbitrary :-) Even I didn't realize that, so thanks for the tip for just testing the first candidate :+1: 

But I removed the comment, might not be of so much value mentioning it there.

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -45,6 +45,13 @@ require('./FlameGraph.css');
 
+/**
+ * How "wide" a call node box needs to be for it to be able to be
+ * selected with keyboard navigation. This is relative to 1, where 1
```
Instead of `This is relative to 1`, maybe `This is a fraction between 0 and 1`.

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -90,17 +97,65 @@ class FlameGraph extends React.PureComponent<Props> {
+   * This means we're already at the end, or the boxes of all
+   * candidate call nodes are too narrow to be selected.
+   */
+  _nextSelectableInRow = (startingCallNodeIndex, direction) => {
```
nit: please type the arguments and the result; especially you can type `direction` like `1 | -1`

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -90,17 +97,65 @@ class FlameGraph extends React.PureComponent<Props> {
+  /**
+   * Is the box for this call node wide enough to be selected?
+   */
+  _wideEnough = callNodeIndex => {
```
nit: please type arguments and the return value

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -90,17 +97,65 @@ class FlameGraph extends React.PureComponent<Props> {
+    while (callNodeIndex !== undefined) {
+      const depth = callNodeTable.depth[callNodeIndex];
+      const row = flameGraphTiming[depth];
+      const rowIndex = row.callNode.indexOf(callNodeIndex) + direction;
```
I believe this is a `columnIndex` and not a `rowIndex` here as well.

julienw replied:
I think you can also look up `row.callNode.indexOf(callNodeIndex)` before the loop, and work from this value instead of look it up again at each step.

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -90,17 +97,65 @@ class FlameGraph extends React.PureComponent<Props> {
+
+    const depth = callNodeTable.depth[callNodeIndex];
+    const row = flameGraphTiming[depth];
+    const rowIndex = row.callNode.indexOf(callNodeIndex);
```
I believe this is a `columnIndex` and not a `rowIndex` here.

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -90,17 +97,65 @@ class FlameGraph extends React.PureComponent<Props> {
+    let callNodeIndex = startingCallNodeIndex;
+    while (callNodeIndex !== undefined) {
+      const depth = callNodeTable.depth[callNodeIndex];
+      const row = flameGraphTiming[depth];
```
I believe the depth and the row are always the same, so you can look them up before the loop

brisad commented:
Thanks, great comments @julienw ! :-) I've addressed them and pushed a new commit.

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -171,13 +181,8 @@ class FlameGraph extends React.PureComponent<Props> {
-        const callNodeIndex = children.find(this._wideEnough);
-
-        if (callNodeIndex !== undefined) {
+        const [callNodeIndex] = callTree.getChildren(selectedCallNodeIndex);
```
nit: still maybe it's worth explaining in a comment that the children are sorted by time length :)

julienw commented on src/components/flame-graph/FlameGraph.js:
```diff
@@ -90,6 +97,120 @@ class FlameGraph extends React.PureComponent<Props> {
+  _nextSelectableInRow = (
+    startingCallNodeIndex: IndexIntoCallNodeTable,
+    direction: 1 | -1
+  ): ?IndexIntoCallNodeTable => {
```
nit: here I think it returns `IndexIntoCallNodeTable | void`; remember that `?IndexIntoCallNodeTable` includes `null` as well.
